### PR TITLE
Homeworks and tests

### PR DESCRIPTION
**Pull request by Shakirov-ER**

### CODE REVIEW

File: module_05/test.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+1. Переменная x имеет тип numpy.array, что вернет x[x > 3] ?
+
+- Массив элементов с индексом больше 3
```
Comment (svkov) [inline module_05/test.md]: мб "элементы массива с индексом больше 3"?

File: module_05/test.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+- Массив элементов с индексом больше 2 
+- Массив элементов  исходного массива, значения которых больше 3, в той же последовательности, в которой они встречались в массиве  **this**
+- Массив элементов массива, значения которых больше 3, в произвольном порядке
+
+2. Выберите все верные утверждения.
+
+- numpy.random.rand(4, 5) вернет двумерный массив размером 4 на 5.  **this**
+- numpy.random.rand(4, 5) вернет числа, сгенерированные из равномерного распределения от 4 до 5.
```
Comment (svkov) [inline module_05/test.md]: мы здесь про равномерное распределение еще не говорили, можно просто сказать "вернет случайное число от 4 до 5"

File: module_05/test.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+- numpy.arange работает существенно быстрее range при генерации большого количества чисел. **this**
```
Comment (svkov) [inline module_05/test.md]: лучше будет "при генерации коллекций большого размера"

File: module_05/test.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+Выберите команды, которые позволяют добавить в frame новый столбец под названием IsStudent так, чтобы после добавления столбца получившийся объект frame выглядел следующим образом:
+
+<table><tbody><tr><td><p></p></td><td><p>Name</p></td><td><p>Birth</p></td><td><p>IsStudent</p></td></tr><tr><td><p>1</p></td><td><p>Михаил Зайцев</p></td><td><p>22.04.1976</p></td><td><p>False</p></td></tr><tr><td><p>2</p></td><td><p>Максим Королёв</p></td><td><p>08.12.1993</p></td><td><p>True</p></td></tr></tbody></table>
+
+Примечание. В python несколько выражений, записанных через запятую и стоящих после "=" или "return", считаются кортежем (tuple). ***
+
+- frame.addColumn('IsStudent', [False, True])
+- frame['IsStudent'] = [False, True]    **this**
+- frame['IsStudent'] = False, True      **this**
+- frame['IsStudent'] = (False, True)    **this**
```
Comment (svkov) [inline module_05/test.md:36]: Можно добавить еще вариант
`frame['IsStudent'] = False and True`

File: module_06/test.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+- имеющих доход не менее 3500, и старше тех, кто родился в 1958 году.
+- имеющих доход менее 3500, ипи тех, кто родился е 1958 году и позже;
+- имеющих доход менее 3500, и родившихся в 1958 году и позже;
+- имеющих доход менее 3500, и родившихся в 1959 году и позже;  **this**
+- имеющих доход менее 3500, и тех, кто родился в 1958 году;
+
+4. Имеются элементы запроса: 
+
+1 SELECT employees.name, departments.name;
```
Comment (svkov) [inline module_06/test.md]: кажется, лучше убрать точки с запятой в конце строки, это выглядит как окончание запроса

File: module_07/test.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+- Скорость роста $f(x,y)$ в точке $(x_0, y_0)$ вдоль оси y
+
+2. Вектор, координатами которого являются частные производные функции в данной точке ***
+
+- Никак не называется, этот объект вообще бесполезен
+- Называется градиентом функции в этой точке **this**
+- Задает направление наискорейшего убывания функции в этой точке
+- Задает направление наискорейшего возрастания функции в этой точке **this**
+
+3. Дана функция трех переменных: $f(x, y, z) = x^2 + y^2 + z^2  и точка (1, 1, 1). Какой шаг должен быть у градиентного спуска, чтобы за одну итерацию достигнуть минимума функции
```
Comment (svkov) [inline module_07/test.md]: что означает "шаг градиентного спуска"? если речь про learning rate, то лучше так и сформулировать

File: module_08/test after 1_2.md [hunk 7]
```diff
@@ -0,0 +1,24 @@
+С помощью каких из приведённых функций можно задавать распределения дискретных случайных величин?  
+
+- Функция распределения.
```
Comment (svkov) [inline module_08/test after 1_2.md:3]: не выбраны правильные ответы

File: module_09/test.md [hunk 8]
```diff
@@ -0,0 +1,23 @@
+1. Пусть $(X\sim F(x))$ — случайная величина с произвольным распределением, неизвестным математическим ожиданием $(\mathbb{E}X)$ и известной дисперсией $(\mathbb{D}X)$. Как выглядит доверительный интервал для $(\mathbb{E}X)$ с приближённым уровнем доверия $(100(1-\alpha))$%?
+
+ - $(\bar{X}_n \pm z_{1-\frac{\alpha}{2}} \frac{\mathbb{D}X}{\sqrt{n}})$
```
Comment (svkov) [inline module_09/test.md]: гитхаб не рендерить эти формулы, надо чекнуть, что катех умеет рендерить такое

File: module_10/test after 3.md [hunk 9]
```diff
@@ -0,0 +1,31 @@
+1. Как можно избавиться от свободного члена в линейных моделях?
+
+- Добавить единичный признак.
```
Comment (svkov) [inline module_10/test after 3.md]: кажется, это правильный ответ

Comment (svkov) [reply module_10/test after 3.md]: но вообще, вопрос немного странный
я бы сформулировал типа "как для упрощения формул избавляются от свободного члена?"

File: module_11/hw.md [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+Построить еще несколько моделей, используя подбор гиперпараметров. Сравнить между собой все построенные модели.
+
+## Hard
+
+Построить модель стекинга над всеми моделями из задачи Normal + добавить один из градиентных бустингов XGBoost/LightGBM/CatBoost, если они не использовались. Удалось ли улучшить результат?
+
+## Hard+
+
+Иногда нужно использовать модели, у которых интерфейс отличается от интерфейса моделей из sklearn и тогда приходится самостоятельно реализовывать стекинг.
```
Comment (svkov) [inline module_11/hw.md]: Нужны пояснение типа такого:

Реализовать модель стекинга, для этого нужно сделать два метода `fit_stacking(x_train, y_train, base_models, final_model)`, который на вход принимает датасет, базовые модели в виде списка и финализирующую модель, и возвращает кортеж из обученных `base_models` и `final_model`. Также понадобится метод `predict_stacking(x_test, base_models, final_model)`, который сделает предсказание по входящим данным.
Оценить качество такой модели, сравнимо ли оно с качеством стекинга из sklearn?

File: module_11/test after 2.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+- Аналитически — можно в явном виде выписать формулы, задающие структуру оптимального дерева.
+
+
+4. Мы решаем задачу классификации с помощью решающего дерева. Ниже приведены разные варианты распределения классов в вершине ([c1, c2, c3] означает, что в вершине c1 объектов первого класса, c2 объектов второго класса и c3 объектов третьего класса). Какой из них должен получить меньшее значение критерия информативности?
+
+- [45, 45, 10]
+- [95, 3, 2]    **this**
+- [33, 34, 33]
+
+5. Какие параметры есть у метода kk ближайших соседей? ***
```
Comment (svkov) [inline module_11/test after 2.md]: очепятка - должна быть одна k

File: module_11/test after 2.md [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+- Число деревьев
+- Число соседей $k$ **this**
+- Функция активации
+- Функция весов объектов    **this**
```
Comment (svkov) [inline module_11/test after 2.md]: про это кажется не было в теории, я бы поменял на метрику для вычисления расстояния

File: module_11/test after 2.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+6. Чем отличаются ответы метода kk ближайших соседей в случае классификации и регрессии?
```
Comment (svkov) [inline module_11/test after 2.md]: опять очепятка

File: module_11/test after 2.md [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+
+- В случае классификации для нового объекта берется наиболее частый ответ среди ответов на kk его ближайших соседей в обучающей выборке, а в случае регрессии — средний ответ.
```
Comment (svkov) [inline module_11/test after 2.md]: не указан правильный ответ

File: module_11/test after 2.md [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+- Ничем. В обоих случаях для нового объекта берется средний ответ среди ответов на kk его ближайших соседей в обучающей выборке.
+
+7. Как выглядит решающее правило для байесовского классификатора?
+
+- $a(x) = \operatorname*{argmax}_y P(y)$
+- $a(x) = \operatorname*{argmax}_y P(x|y)$
+- $a(x) = \operatorname*{argmax}_y P(x|y)P(y)$  **this**
+- $a(x) = \operatorname*{argmax}_y P(x)$
+
+8. Пусть у нас есть $k$ классов, и априорные вероятности для всех них равны $P(y) = \dfrac{1}{k}$P. Как тогда выглядит решающее правило байесовского классификатора?
```
Comment (svkov) [inline module_11/test after 2.md]: формула не рендерится гитхабом, но кажется там лишняя  P в конце

File: module_13/test.md [hunk 16]
```diff
@@ -0,0 +1,29 @@
+1. Временной ряд это:
+- произвольно расположенные в произвольном порядке показатели, характеризующие развитие явления во времени;
+- последовательно расположенные в хронологическом порядке показатели, с постоянной разницей во времени, характеризующие развитие явления во времени;  **this**
+- последовательно расположенные по мере возрастания показатели, характеризующие развитие явления во времени.
+
+2. Временной ряд называется стационарным, если:
+
+- среднее значение членов ряда постоянно    **this**
```
Comment (svkov) [inline module_13/test.md]: немного не так
матожидание, дисперсия и ковариация постоянны

File: module_14/hw.md [hunk 17]
```diff
@@ -0,0 +1,21 @@
+# Домашняя работа
+
+Будем решать задачу сентимент-анализа. В архиве `data/sentiment analisys.zip` лежит несколько эксель файлов с твитами разных пользователей на разных языках (один файл - один язык).
```
Comment (svkov) [inline module_14/hw.md:3]: а где этот самый архив? надо пояснить, где его найти

File: module_14/hw.md [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+
+Оценить сбалансированность классов.
+
+Удалить пунктуацию из датасета.
+
+Преобразовать датасет в BOW или TF-IDF, поделить на треин-тест и спрогнозировать вашей любимой моделью
+
+## Normal
+
+Удалить стоп-слова из датасета, использовать стеммер Портера (лежит в `src/porter.py`). Использовать пайплайн для TF-IDF и модели машинного обучения. Подобрать параметры. Сравнить качество нескольких моделей.
```
Comment (svkov) [inline module_14/hw.md]: тут тоже надо что-то придумать со стеммером, вроде я в теории оставлял ссылку на исходники, тогда можно просто убрать про путь до стеммера

File: module_15/test.md [hunk 19]
```diff
@@ -0,0 +1,31 @@
+1. Что является входом искусственного нейрона?
+
+-  множество сигналов **this**
+-  единственный сигнал 
+-  весовые значения 
+-  значения активационной функции
+
+2. Активационной функцией называется:
+
+- функция, вычисляющая выходной сигнал нейрона
```
Comment (svkov) [inline module_15/test.md]: не указан правильный ответ

File: module_16/hw.md [hunk 20]
```diff
@@ -0,0 +1,13 @@
+## Домашнее задание
```
Comment (svkov) [inline module_16/hw.md:1]: нужен какой-то контекст, какие-то вводные слова, типа на уроке вы начали осваивать snakemake и docker, в качестве домашнего задания попробуйте улучшить полученное решение

а еще тут надо подумать как это потом проверяться будет, потому что до этого просто достаточно юпитер ноутбука было, а здесь в юпитер не засунешь это

**Review by svkov — CHANGES_REQUESTED** [review]
в основном все ок, по формулировкам где-то поправил только

нужно только подумать про две домашки:
- 14 модуль - я оставлял ссылку на репозиторий с оффлайн-курса, нужно просто прикрепить ссылку на файл, думаю этого будет достаточно
- 16 модуль - тут подумать как принимать домашку, может явно попросить их создать новый репо для этой домашки, либо в основном репозитории и докерфайл и снейкфайл положить в папку с домашкой, нужно просто прописать это явно, чтобы все одинаково сделали

**Comment by svkov** [comment]
Закрывает #20 #21

File: module_09/test.md (hunk 8 above)
Comment (Shakirov-ER) [reply module_09/test.md]: Чекни плз, у меня на компе все рендерится

File: module_10/test after 3.md (hunk 9 above)
Comment (Shakirov-ER) [reply module_10/test after 3.md]: Да, поставил правильный ответ не туда))

File: module_05/test.md (hunk 1 above)
Comment (Shakirov-ER) [reply module_05/test.md]: done

File: module_05/test.md (hunk 2 above)
Comment (Shakirov-ER) [reply module_05/test.md]: done

File: module_05/test.md (hunk 3 above)
Comment (Shakirov-ER) [reply module_05/test.md]: done

File: module_05/test.md (hunk 4 above)
Comment (Shakirov-ER) [reply module_05/test.md:36]: done

File: module_06/test.md (hunk 5 above)
Comment (Shakirov-ER) [reply module_06/test.md]: done

File: module_07/test.md (hunk 6 above)
Comment (Shakirov-ER) [reply module_07/test.md]: done

File: module_08/test after 1_2.md (hunk 7 above)
Comment (Shakirov-ER) [reply module_08/test after 1_2.md:3]: done

File: module_10/test after 3.md (hunk 9 above)
Comment (Shakirov-ER) [reply module_10/test after 3.md]: done

File: module_11/hw.md (hunk 10 above)
Comment (Shakirov-ER) [reply module_11/hw.md]: done 

(Тут лукс лайк не вставил часть текста)

File: module_11/test after 2.md (hunk 11 above)
Comment (Shakirov-ER) [reply module_11/test after 2.md]: done

File: module_11/test after 2.md (hunk 12 above)
Comment (Shakirov-ER) [reply module_11/test after 2.md]: done

File: module_11/test after 2.md (hunk 13 above)
Comment (Shakirov-ER) [reply module_11/test after 2.md]: done

File: module_11/test after 2.md (hunk 14 above)
Comment (Shakirov-ER) [reply module_11/test after 2.md]: done

File: module_11/test after 2.md (hunk 15 above)
Comment (Shakirov-ER) [reply module_11/test after 2.md]: done

File: module_13/test.md (hunk 16 above)
Comment (Shakirov-ER) [reply module_13/test.md]: done

File: module_14/hw.md (hunk 18 above)
Comment (Shakirov-ER) [reply module_14/hw.md]: done

File: module_15/test.md (hunk 19 above)
Comment (Shakirov-ER) [reply module_15/test.md]: done

File: module_16/hw.md (hunk 20 above)
Comment (Shakirov-ER) [reply module_16/hw.md:1]: done
